Web Inspector: Crash in Inspector::InspectorDebuggerAgent::didScheduleAsyncCall on knowyourmeme.com

robert-jenner opened the pull request:
#### 6a341af34a111bdf2a2fd229ead679f7b373b3eb
<pre>
Web Inspector: Crash in Inspector::InspectorDebuggerAgent::didScheduleAsyncCall on knowyourmeme.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=270139">https://bugs.webkit.org/show_bug.cgi?id=270139</a>
<a href="https://rdar.apple.com/123667952">rdar://123667952</a>

Reviewed by Pascoe.

Avoid reading past the end of `m_pendingAsyncCalls`.

* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::didScheduleAsyncCall):

Originally-landed-as: 272448.684@safari-7618-branch (cd62341b1dda). <a href="https://rdar.apple.com/128090099">rdar://128090099</a>
Canonical link: <a href="https://commits.webkit.org/278872@main">https://commits.webkit.org/278872@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c198e346af651232039cdcbf89748f17c850544

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4049 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54953 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2379 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37364 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2063 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42069 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28640 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23199 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25936 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1847 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45029 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47896 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1938 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56545 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51192 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1813 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49470 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28045 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44657 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48697 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/is-playing-audio (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28942 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/63512 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7566 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27782 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11986 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
```diff
@@ -444,7 +444,8 @@ void InspectorDebuggerAgent::didScheduleAsyncCall(JSC::JSGlobalObject* globalObj
         auto it = m_pendingAsyncCalls.find(m_currentAsyncCallIdentifierStack.last());
         ASSERT(it != m_pendingAsyncCalls.end());
-        parentStackTrace = it->value;
+        if (LIKELY(it != m_pendingAsyncCalls.end()))
```
do we actually know whether this is expected/allowed?  or is this scenario a bug?

if the former, we should remove the `ASSERT` just above (and also probably the `UNLIKELY` too) since it's no longer considered impossible

if the latter, do we know why this happens?  perhaps we should also be checking something like this too?
```diff
-ASSERT(!m_currentAsyncCallIdentifierStack.contains(identifier));
-
-if (!asyncStackTrace->isPending())
+if (!asyncStackTrace->isPending() && !m_currentAsyncCallIdentifierStack.contains(identifier))
     m_pendingAsyncCalls.remove(identifier);
```
i.e. i wonder if any of the `ASSERT` involving `m_currentAsyncCallIdentifierStack` are actually valid

robert-jenner replied:
Excellent question @dcrousso this is a merge-back commit. So I will ask @charliewolfe care to comment?

charliewolfe replied:
It wasn’t obvious to me if this was a valid state. The original author seemed to think not, so I left the assert. Was just trying to avoid the OOB read here.

dcrousso replied:
wait hold on what?

this was added 8yrs ago in [184626@main](https://commits.webkit.org/184626@main).  the original author hasn't worked at Apple for many years.  did you reach out to them externally or something?

the last person to work on this stuff was me in [252543@main](https://commits.webkit.org/252543@main) and [252869@main](https://commits.webkit.org/252869@main)

i dont think this change is necessarily wrong, but im concerned it's covering up a larger issue (e.g. somehow `m_currentAsyncCallIdentifierStack` and `m_pendingAsyncCalls` are getting out of sync) which i think merits further investigation

charliewolfe replied:
No - I didn't reach out to them, I just didn't have good reason to believe this assert was now invalid. I'm sure you're correct there is a larger bug here that should be investigated further.

charliewolfe committed:
Web Inspector: Crash in Inspector::InspectorDebuggerAgent::didScheduleAsyncCall on knowyourmeme.com
https://bugs.webkit.org/show_bug.cgi?id=270139
rdar://123667952

Reviewed by Pascoe.

Avoid reading past the end of `m_pendingAsyncCalls`.

* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::didScheduleAsyncCall):

Originally-landed-as: 272448.684@safari-7618-branch (cd62341b1dda). rdar://128090099
Canonical link: https://commits.webkit.org/278872@main
